processing/src/main/java/io/druid/query/Query.java

```diff
@@ -90,13 +98,25 @@
 
   Ordering<T> getResultOrdering();
 
-  Query<T> withOverriddenContext(Map<String, Object> contextOverride);
-
-  Query<T> withQuerySegmentSpec(QuerySegmentSpec spec);
+  String getId();
 
   Query<T> withId(String id);
 
-  String getId();
+  default DataSourceWithSegmentSpec getDistributionTarget()
+  {
+    return getContextValue(QueryContextKeys.DIST_TARGET_SOURCE);
+  }
+
+  default Query<T> distributeBy(DataSourceWithSegmentSpec spec)
+  {
+    return withOverriddenContext(ImmutableMap.of(QueryContextKeys.DIST_TARGET_SOURCE, spec));
+  }
+
+  Query<T> withOverriddenContext(Map<String, Object> contextOverride);
+
+  Query<T> replaceQuerySegmentSpecWith(DataSource dataSource, QuerySegmentSpec spec);
```

Comment:
I think this method should follow `withXxx` convention, i. e. "withQuerySegmentSpec()"